Add ability to specify system namespace during cluster add operation

jannfis opened the pull request:
This change addresses #1595 by adding the --system-namespace option to the "cluster add" command of argocd CLI as well as a corresponding field in the API function:

[eris@docker-build argo-cd]$ ./dist/argocd cluster add --help | grep namespace
      --system-namespace string   Use different system namespace (default "kube-system")

When given, ArgoCD will create the System Account "argocd-manager" in the specified namespace. The change is non-breaking, defaults to previous behaviour by using "kube-system" as namespace when the option was not specified.

jannfis commented:
Again, this passes both "test" and "test-e2e" on my workstation and Argo CI throws the following error at the E2E stage:

time="2019-05-29T17:00:52Z" level=info msg="ran command" duration=80.687149ms err="exit status 1"                                                                                                                                                     
time="2019-05-29T17:00:52Z" level=info msg="running command" args="[app get argocd-e2e-zvauz -o yaml --server 10.43.142.3:443 --auth-token eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpYXQiOjE1NTkxNDkyNTEsImlzcyI6ImFyZ29jZCIsIm5iZiI6MTU1OTE0OTI1MSwic3
ViIjoiYWRtaW4ifQ.Kh33s0H_3DJXsuVqDHelscECH2kNwXVCnqqvKBfkd-c --insecure]" name=../../dist/argocd workDir=                                                                                                                                             
time="2019-05-29T17:00:52Z" level=info msg="0: time=\"2019-05-29T17:00:52Z\" level=fatal msg=\"rpc error: code = Unauthenticated desc = invalid session: signature is invalid\""

alexec commented on server/cluster/cluster.proto:
```diff
@@ -29,6 +29,7 @@ message ClusterCreateFromKubeConfigRequest {
 	string context = 2;
 	bool upsert = 3;
 	bool inCluster = 4;
+	string systemNamespace = 5;
```
This is part of the public API of Argo CD and therefore for needs to be documented (yes, I know the other values aren't).

Can you please add a comment to this the run `make codegen`?  It'll then update `swagger.json` automatically.

Otherwise looks good to me.

jannfis replied:
Fixed with latest commit.

jannfis commented:
Hi @alexec 

I have done both - merging from master and make pre-commit - already before submitting the PR. Actually, I branched off from latest upstream master, then pulled in the relevant changes from a previous local branch selectively. I'm still rather new to collaborating on GitHub (and also to working with git as a tool for that matter). What is the expected outcome of doing it again? 

Running make protogen also re-generated some of the other files, which got me quite confused - I didn't include them in the PR, because I neither touched their sources nor do I understand the changes yet:

        modified:   pkg/apis/application/v1alpha1/generated.pb.go
        modified:   reposerver/repository/repository.pb.go
        modified:   server/account/account.pb.go
        modified:   server/application/application.pb.go
        modified:   server/project/project.pb.go
        modified:   server/repository/repository.pb.go
        modified:   server/session/session.pb.go
        modified:   server/settings/settings.pb.go
        modified:   server/version/version.pb.go

Shall I push them to the PR as well?

Regarding listing my organization in the README - it's not that simple, unfortunately... I work with a fairly huge and very public corporation which has certain processes and regulations in place before having their name displayed in a public place. I already initiated a request for reviewing whether that's possible - because I'd love to see us in there, too :) Anyway, all my code contributions are created in my free time - I've got no coder job in the corp (read: no time for coding during work hours), but I'm working towards the broad adoption of ArgoCD for our K8S-based projects.

alexec commented:
Hmmmmmmm... do you want to write tests for this?

jannfis commented:
Sure, let me have a look at how the tests work - I guess the right place for this would be in the end-to-end test suite? As far as I understood up until now, the end-to-end tests work without the ArgoCD components running in a k8s cluster but running local. 

I'll dig my way in, any pointers are appreciated!

alexec commented:
I think you should be able to write unit tests for this. I don't think E2E tests support different namespaces, but please prove me wrong!

https://argoproj.github.io/argo-cd/developer-guide/test-e2e/

jannfis commented:
I seem to not get my head around this. Both, unit and e2e tests do not work with an installation of ArgoCD running in a K8S cluster. 

This change however affects integration of a k8s cluster with ArgoCD, and only at the time when we incorporate a remote k8s cluster into ArgoCD configuration for the first time. 

Not saying a test would not be important, but I cannot seem to find a valid test-case for this change in the current construct :-(

alexec commented:
If there’s no straight forward valid test, we can probably skip it.

alexec commented:
Note, I think you still need to sync with master.

jannfis commented:
Shall I commit above mentioned protobuf generated files as well?

alexec commented:
hi - the `codegen` build checks you've generated the code right, can you double check - you may have different versions of the `protoc` tool installed (should be 3.7.1)

alexec commented:
I've merged so we can move forward.

jannfis commented:
@alexec Thanks and sorry for the delay in my answer, I was away for some days.

Checked the version of my protoc toolchain, guess I'm lagging a bit behind (3.6.1). Will upgrade for the next PRs, thank you for the hint.
